src/main/java/com/shapesecurity/shift/es2017/parser/PatternAcceptor.java

```diff
@@ -0,0 +1,721 @@
+        }
+
+        public State backtrackOnFailure() {
+            return new State(this);
+        }
+
+        public boolean backtrackOnFailure(F<State, Boolean> predicate) {
+            State state = this.backtrackOnFailure();
+            boolean accepted = predicate.apply(state);
+            if (accepted) {
+                this.absorb(state);
+            }
+            return accepted;
+        }
+
+        public <B> Maybe<B> backtrackOnFailureMaybe(F<State, Maybe<B>> predicate) {
+            State state = this.backtrackOnFailure();
+            Maybe<B> accepted = predicate.apply(state);
+            if (accepted.isJust()) {
+                this.absorb(state);
+            }
+            return accepted;
+        }
+
+        private void absorb(State otherState) {
+            this.index = otherState.index;
+            this.backreferenceNames = otherState.backreferenceNames;
+            this.largestBackreference = otherState.largestBackreference;
+            this.groupingNames = otherState.groupingNames;
+            this.capturingGroups = otherState.capturingGroups;
+        }
+
+        public Maybe<String> nextCodePoint() {
+            if (this.index >= pattern.length()) {
+                return Maybe.empty();
+            }
+            if (!PatternAcceptor.this.unicode) {
+                return Maybe.of(new String(Character.toChars(pattern.charAt(this.index))));
+            }
+            return Maybe.of(new String(Character.toChars(pattern.codePointAt(this.index))));
+        }
+
+        public void skipCodePoint() {
+            this.nextCodePoint().foreach(cp -> this.index += cp.length());
+        }
+
+        public boolean eat(String str) {
+            if (this.index + str.length() > pattern.length() || !pattern.startsWith(str, this.index)) {
+                return false;
+            }
+            this.index += str.length();
+            return true;
+        }
+
+        public Maybe<String> eatAny(@Nonnull String... strings) {
+            for (String string : strings) {
+                if (this.eat(string)) {
+                    return Maybe.of(string);
+                }
+            }
+            return Maybe.empty();
+        }
+
+        public Maybe<String> eatAny(@Nonnull String[]... stringArrays) {
+            for (String[] strings : stringArrays) {
+                for (String string : strings) {
+                    if (this.eat(string)) {
+                        return Maybe.of(string);
+                    }
+                }
+            }
+            return Maybe.empty();
+        }
+
+        public String collect(@Nonnull String[]... stringArrays) {
+            return collect(Maybe.empty(), stringArrays);
+        }
+
+        public String collect(Maybe<Integer> limit, @Nonnull String[]... stringArrays) {
+            StringBuilder stringBuilder = new StringBuilder();
+
+            outer:
+            for (int i = 0; limit.isNothing() || limit.fromJust() > i; ++i) {
```

Comment:
Let's pull out the `isNothing` and `fromJust` from the test, so it doesn't evaluate every time.

Reply:
I see I am not the only one with no faith in the JVM. :D